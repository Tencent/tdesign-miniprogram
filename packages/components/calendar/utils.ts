export function getMonthByOffset(date: Date, offset: number) {
  const _date = new Date(date);
  _date.setMonth(_date.getMonth() + offset);
  _date.setDate(1);
  return _date;
}

export function getYearByOffset(date: Date, offset: number) {
  const _date = new Date(date);
  _date.setFullYear(_date.getFullYear() + offset);
  _date.setDate(1);
  return _date;
}

export const getPrevMonth = (date: Date) => getMonthByOffset(date, -1);
export const getNextMonth = (date: Date) => getMonthByOffset(date, 1);
export const getPrevYear = (date: Date) => getYearByOffset(date, -1);
export const getNextYear = (date: Date) => getYearByOffset(date, 1);
