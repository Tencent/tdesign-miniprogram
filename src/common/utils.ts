import { prefix } from './config';

const systemInfo = wx.getSystemInfoSync();

type Context = WechatMiniprogram.Page.TrivialInstance | WechatMiniprogram.Component.TrivialInstance;

export const debounce = function (func, wait = 500) {
  let timerId;
  return function (...rest) {
    if (timerId) {
      clearTimeout(timerId);
    }
    timerId = setTimeout(() => {
      func.apply(this, rest);
    }, wait);
  };
};

export const throttle = (func, wait = 100, options = null) => {
  let previous = 0;
  let timerid = null;

  if (!options) {
    options = {
      leading: true,
    };
  }

  return function (...args) {
    const now = Date.now();

    if (!previous && !options.leading) previous = now;

    const remaining = wait - (now - previous);
    const context = this;

    if (remaining <= 0) {
      if (timerid) {
        clearTimeout(timerid);
        timerid = null;
      }
      previous = now;
      func.apply(context, args);
    }
  };
};

export const classNames = function (...args) {
  const hasOwn = {}.hasOwnProperty;
  const classes = [];

  args.forEach((arg) => {
    // for (let i = 0; i < args.length; i++) {
    // eslint-disable-next-line
    // const arg = args[i]
    if (!arg) return;

    const argType = typeof arg;

    if (argType === 'string' || argType === 'number') {
      classes.push(arg);
    } else if (Array.isArray(arg) && arg.length) {
      const inner = classNames(...arg);
      if (inner) {
        classes.push(inner);
      }
    } else if (argType === 'object') {
      // eslint-disable-next-line
      for (const key in arg) {
        if (hasOwn.call(arg, key) && arg[key]) {
          classes.push(key);
        }
      }
    }
  });

  return classes.join(' ');
};

export const styles = function (styleObj) {
  return Object.keys(styleObj)
    .map((styleKey) => `${styleKey}: ${styleObj[styleKey]}`)
    .join('; ');
};

export const getAnimationFrame = function (context: any, cb: Function) {
  return wx
    .createSelectorQuery()
    .in(context)
    .selectViewport()
    .boundingClientRect()
    .exec(() => {
      cb();
    });
};

export const getRect = function (context: any, selector: string, needAll: boolean = false) {
  return new Promise<any>((resolve, reject) => {
    wx.createSelectorQuery()
      .in(context)
      [needAll ? 'selectAll' : 'select'](selector)
      .boundingClientRect((rect) => {
        if (rect) {
          resolve(rect);
        } else {
          reject(rect);
        }
      })
      .exec();
  });
};

const isDef = function (value: any): boolean {
  return value !== undefined && value !== null;
};

export const isNumber = function (value) {
  return /^\d+(\.\d+)?$/.test(value);
};

export const addUnit = function (value?: string | number): string | undefined {
  if (!isDef(value)) {
    return undefined;
  }
  value = String(value);
  return isNumber(value) ? `${value}px` : value;
};

/**
 * 计算字符串字符的长度并可以截取字符串。
 * @param str 传入字符串（maxcharacter条件下，一个汉字表示两个字符）
 * @param max 规定最大字符串长度
 * @returns 当没有传入maxCharacter/maxLength 时返回字符串字符长度，当传入maxCharacter/maxLength时返回截取之后的字符串和长度。
 */
export const getCharacterLength = (type: string, str: string, max?: number) => {
  if (!str || str.length === 0) {
    return {
      length: 0,
      characters: '',
    };
  }

  if (type === 'maxcharacter') {
    let len = 0;
    for (let i = 0; i < str.length; i += 1) {
      let currentStringLength = 0;
      if (str.charCodeAt(i) > 127 || str.charCodeAt(i) === 94) {
        currentStringLength = 2;
      } else {
        currentStringLength = 1;
      }
      if (len + currentStringLength > max) {
        return {
          length: len,
          characters: str.slice(0, i),
        };
      }
      len += currentStringLength;
    }
    return {
      length: len,
      characters: str,
    };
  } else if (type === 'maxlength') {
    const length = str.length > max ? max : str.length;
    return {
      length,
      characters: str.slice(0, length),
    };
  }

  return {
    length: str.length,
    characters: str,
  };
};

export const chunk = (arr: any[], size: number) =>
  Array.from({ length: Math.ceil(arr.length / size) }, (v, i) => arr.slice(i * size, i * size + size));

export const getInstance = function (context?: Context, selector?: string) {
  if (!context) {
    const pages = getCurrentPages();
    const page = pages[pages.length - 1];
    context = page.$$basePage || page;
  }
  const instance = context ? context.selectComponent(selector) : null;
  if (!instance) {
    console.warn('未找到组件,请检查selector是否正确');
    return null;
  }
  return instance;
};

export const unitConvert = (value: number | string): number => {
  if (typeof value === 'string') {
    if (value.includes('rpx')) {
      return (parseInt(value, 10) * (systemInfo?.screenWidth ?? 750)) / 750;
    }
    return parseInt(value, 10);
  }
  return value;
};

export const setIcon = (iconName, icon, defaultIcon) => {
  if (icon) {
    if (typeof icon === 'string') {
      return {
        [`${iconName}Name`]: icon,
        [`${iconName}Data`]: {},
      };
    } else if (typeof icon === 'object') {
      return {
        [`${iconName}Name`]: '',
        [`${iconName}Data`]: icon,
      };
    } else {
      return {
        [`${iconName}Name`]: defaultIcon,
        [`${iconName}Data`]: {},
      };
    }
  }
  return {
    [`${iconName}Name`]: '',
    [`${iconName}Data`]: {},
  };
};

export const isBool = (val) => typeof val === 'boolean';

export const isObject = (val) => typeof val === 'object' && val != null;

export const isString = (val) => typeof val === 'string';

export const toCamel = (str) => str.replace(/-(\w)/g, (match, m1) => m1.toUpperCase());

export const getCurrentPage = function <T>() {
  const pages = getCurrentPages();
  return pages[pages.length - 1] as T & WechatMiniprogram.Page.TrivialInstance;
};

export const uniqueFactory = (compName) => {
  let number = 0;
  return () => `${prefix}_${compName}_${number++}`;
};

export const calcIcon = (icon: string | Record<string, any>, defaultIcon?: string) => {
  if ((isBool(icon) && icon && defaultIcon) || isString(icon)) {
    return { name: isBool(icon) ? defaultIcon : icon };
  }
  if (isObject(icon)) {
    return icon;
  }
  return null;
};

export const isOverSize = (size, sizeLimit) => {
  if (!sizeLimit) return false;

  const base = 1000;
  const unitMap = {
    B: 1,
    KB: base,
    MB: base * base,
    GB: base * base * base,
  };
  const computedSize =
    typeof sizeLimit === 'number' ? sizeLimit * base : sizeLimit?.size * unitMap[sizeLimit?.unit ?? 'KB']; // 单位 KB

  return size > computedSize;
};
