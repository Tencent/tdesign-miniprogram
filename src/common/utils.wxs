/* utils */

/**
 * addUnit */
// 为 css 添加单位
function addUnit(value) {
  var REGEXP = getRegExp('^-?\d+(.\d+)?$');
  if (value == null) {
    return undefined;
  }
  return REGEXP.test('' + value) ? value + 'px' : value;
}

function isString(string) {
  return string && string.constructor === 'String';
}

function isArray(array) {
  return array && array.constructor === 'Array';
}

function isObject(obj) {
  return obj && obj.constructor === 'Object';
}

var isNoEmptyObj = function (obj) {
  return isObject(obj) && JSON.stringify(obj) !== '{}';
};

function includes(arr, value) {
  if (!arr || !isArray(arr)) return false;

  var i = 0;
  var len = arr.length;

  for (; i < len; i++) {
    if (arr[i] === value) return true;
  }
  return false;
}

function cls(base, arr) {
  var res = [base];
  var i = 0;
  for (var size = arr.length; i < size; i++) {
    var item = arr[i];

    if (item && item.constructor === 'Array') {
      var key = arr[i][0];
      var value = arr[i][1];

      if (value) {
        res.push(base + '--' + key);
      }
    } else if (typeof item === 'string' || typeof item === 'number') {
      if (item) {
        res.push(base + '--' + item);
      }
    }
  }
  return res.join(' ');
}

function getBadgeAriaLabel(options) {
  var maxCount = options.maxCount || 99;
  if (options.dot) {
    return '有新的消息';
  }
  if (options.count === '...') {
    return '有很多消息';
  }
  if (isNaN(options.count)) {
    return options.count;
  }
  var str1 = '有' + maxCount + '+条消息';
  var str2 = '有' + options.count + '条消息';
  return Number(options.count) > maxCount ? str1 : str2;
}

function endsWith(str, endStr) {
  return str.slice(-endStr.length) === endStr ? str : str + endStr;
}

function keys(obj) {
  return JSON.stringify(obj)
    .replace(getRegExp('{|}|"', 'g'), '')
    .split(',')
    .map(function (item) {
      return item.split(':')[0];
    });
}

function kebabCase(str) {
  return str
    .replace(getRegExp('[A-Z]', 'g'), function (ele) {
      return '-' + ele;
    })
    .toLowerCase();
}

function _style(styles) {
  if (isArray(styles)) {
    return styles
      .filter(function (item) {
        return item != null && item !== '';
      })
      .map(function (item) {
        return isArray(item) ? _style(item) : endsWith(item, ';');
      })
      .join(' ');
  }

  if (isObject(styles)) {
    return keys(styles)
      .filter(function (key) {
        return styles[key] != null && styles[key] !== '';
      })
      .map(function (key) {
        return [kebabCase(key), [styles[key]]].join(':');
      })
      .join(';');
  }

  return styles;
}

module.exports = {
  addUnit: addUnit,
  isString: isString,
  isArray: isArray,
  isObject: isObject,
  isNoEmptyObj: isNoEmptyObj,
  includes: includes,
  cls: cls,
  getBadgeAriaLabel: getBadgeAriaLabel,
  _style: _style,
};
