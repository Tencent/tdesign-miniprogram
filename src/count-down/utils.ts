export interface TimeData {
  DD: number;
  HH: number;
  mm: number;
  ss: number;
  SSS: number;
}
export const TimeDataUnit = {
  DD: '天',
  HH: '时',
  mm: '分',
  ss: '秒',
  SSS: '毫秒',
};

const SECOND = 1000;
const MINUTE = 60 * SECOND;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;

export const parseTimeData = function (time: number): TimeData {
  const days = Math.floor(time / DAY);
  const hours = Math.floor((time % DAY) / HOUR);
  const minutes = Math.floor((time % HOUR) / MINUTE);
  const seconds = Math.floor((time % MINUTE) / SECOND);
  const milliseconds = Math.floor(time % SECOND);
  return {
    DD: days,
    HH: hours,
    mm: minutes,
    ss: seconds,
    SSS: milliseconds,
  };
};

export const isSameSecond = function (time1: number, time2: number): boolean {
  return Math.floor(time1 / 1000) === Math.floor(time2 / 1000);
};

export type TTimeList = { digit: string; unit: string; match: string }[];

/**
 *
 * @param time 倒计时时间，毫秒单位
 * @param format 倒计时格式化字符串，例如：dd天hh小时mm分ss秒SSS毫秒，hh:mm:ss.SSS，hh:mm:ss
 */
export const parseFormat = function (time: number, format: string) {
  const obj = {
    'D+': Math.floor(time / 86400000), // 日
    'H+': Math.floor((time % 86400000) / 3600000), // 小时
    'm+': Math.floor((time % 3600000) / 60000), // 分
    's+': Math.floor((time % 60000) / 1000), // 秒
    'S+': Math.floor(time % 1000), // 毫秒
  };
  const timeList: TTimeList = [];
  let timeText = format;

  Object.keys(obj).forEach((prop) => {
    if (new RegExp(`(${prop})`).test(timeText)) {
      timeText = timeText.replace(RegExp.$1, (match, offset, source) => {
        const v = `${(obj as any)[prop]}`;
        let digit = v;
        if (match.length > 1) {
          digit = (match.replace(new RegExp(match[0], 'g'), '0') + v).substr(v.length);
        }
        const unit = source.substr(offset + match.length);
        const last = timeList[timeList.length - 1];
        if (last) {
          const index = last.unit.indexOf(match);
          if (index !== -1) {
            last.unit = last.unit.substr(0, index);
          }
        }
        timeList.push({ digit, unit, match });
        return digit;
      });
    }
  });
  return { timeText, timeList };
};
