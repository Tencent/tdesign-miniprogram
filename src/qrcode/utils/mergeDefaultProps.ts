export function mergeDefaultProps<T extends Record<string, any>>(
  originalProps: T,
  defaultProps: Partial<T>
): T {
  const result: any = { ...originalProps };
  
  for (const key in defaultProps) {
    if (Object.prototype.hasOwnProperty.call(defaultProps, key)) {
      // 只有当原始props中该属性为undefined时，才使用默认值
      if (result[key] === undefined) {
        result[key] = defaultProps[key];
      }
    }
  }
  
  return result as T;
}