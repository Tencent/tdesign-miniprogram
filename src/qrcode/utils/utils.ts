import type {
  ERROR_LEVEL_MAPPED_TYPE,
  ErrorCorrectionLevel,
  Excavation,
  ImageSettings,
  Modules,
} from './type';
import { Ecc } from './qrcodegen';

// =================== ERROR_LEVEL ==========================
export const ERROR_LEVEL_MAP: ERROR_LEVEL_MAPPED_TYPE = {
  L: Ecc.LOW,
  M: Ecc.MEDIUM,
  Q: Ecc.QUARTILE,
  H: Ecc.HIGH,
} as const;

// =================== DEFAULT_VALUE ==========================
export const DEFAULT_SIZE = 160;
export const DEFAULT_LEVEL: ErrorCorrectionLevel = 'M';
export const DEFAULT_BACKGROUND_COLOR = '#FFFFFF';
export const DEFAULT_FRONT_COLOR = '#000000';
export const DEFAULT_NEED_MARGIN = true;
export const DEFAULT_MINVERSION = 1;
export const SPEC_MARGIN_SIZE = 0;
export const DEFAULT_MARGIN_SIZE = 6; // 不同大小的二维码要修改默认margin
export const DEFAULT_IMG_SCALE = 0.1;

// =================== UTILS ==========================
/**
 * Generate a path string from modules
 * @param modules
 * @param margin
 * @returns
 */
export const generatePath = (modules: Modules, margin: number = 0) => {
  const ops: string[] = [];
  modules.forEach((row, y) => {
    let start: number | null = null;
    row.forEach((cell, x) => {
      if (!cell && start !== null) {
        ops.push(
          `M${start + margin} ${y + margin}h${x - start}v1H${start + margin}z`
        );
        start = null;
        return;
      }

      if (x === row.length - 1) {
        if (!cell) {
          return;
        }
        if (start === null) {
          ops.push(`M${x + margin},${y + margin} h1v1H${x + margin}z`);
        } else {
          ops.push(
            `M${start + margin},${y + margin} h${x + 1 - start}v1H${
              start + margin
            }z`
          );
        }
        return;
      }

      if (cell && start === null) {
        start = x;
      }
    });
  });
  return ops.join('');
};

/**
 * Excavate modules
 * @param modules
 * @param excavation
 * @returns
 */
export const excavateModules = (modules: Modules, excavation: Excavation) => modules.slice().map((row: any, y: any) => {
  if (y < excavation.y || y >= excavation.y + excavation.h) {
    return row;
  }
  return row.map((cell: any, x: any) => {
    if (x < excavation.x || x >= excavation.x + excavation.w) {
      return cell;
    }
    return false;
  });
});

/**
 * Get image settings
 * @param cells The modules of the QR code
 * @param size The size of the QR code
 * @param margin
 * @param imageSettings
 * @returns
 */
export const getImageSettings = (
  cells: Modules,
  size: number,
  margin: number,
  imageSettings?: ImageSettings
): null | {
  x: number;
  y: number;
  h: number;
  w: number;
  excavation: Excavation | null;
  opacity: number;
} => {
  if (imageSettings == null) {
    return null;
  }
  const numCells = cells.length + margin * 2;
  const defaultSize = Math.floor(size * DEFAULT_IMG_SCALE);
  const scale = numCells / size;
  const w = (imageSettings.width || defaultSize) * scale;
  const h = (imageSettings.height || defaultSize) * scale;
  const x = imageSettings.x == null
    ? cells.length / 2 - w / 2
    : imageSettings.x * scale;
  const y = imageSettings.y == null
    ? cells.length / 2 - h / 2
    : imageSettings.y * scale;
  const opacity = imageSettings.opacity == null ? 1 : imageSettings.opacity;

  let excavation = null;
  if (imageSettings.excavate) {
    const floorX = Math.floor(x);
    const floorY = Math.floor(y);
    const ceilW = Math.ceil(w + x - floorX);
    const ceilH = Math.ceil(h + y - floorY);
    excavation = { x: floorX, y: floorY, w: ceilW, h: ceilH };
  }

  return { x, y, h, w, excavation, opacity };
};

/**
 * Get margin size
 * @param needMargin Whether need margin
 * @param marginSize Custom margin size
 * @returns
 */
export const getMarginSize = (needMargin: boolean, marginSize?: number) => {
  if (marginSize != null) {
    return Math.max(Math.floor(marginSize), 0);
  }
  return needMargin ? SPEC_MARGIN_SIZE : DEFAULT_MARGIN_SIZE;
};

/**
 * Check if Path2D is supported
 */
export const isSupportPath2d = (() => {
  return false;
})();
