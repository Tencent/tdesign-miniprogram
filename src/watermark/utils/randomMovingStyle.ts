/* eslint-disable no-nested-ternary */
export default function randomMovingStyle() {
  const align = Math.floor(Math.random() * 4);
  const p1 = Math.floor(Math.random() * 70) + 30;
  const leftTopLimit = 0;
  const bottomLimit = 95;
  const rightLimit = 90;
  const left0 = align === 1 ? rightLimit : align === 3 ? leftTopLimit : p1;
  const left1 = align === 0 ? rightLimit : align === 2 ? leftTopLimit : 100 - p1;
  const left2 = align === 1 ? leftTopLimit : align === 3 ? rightLimit : 100 - p1;
  const left3 = align === 0 ? leftTopLimit : align === 2 ? rightLimit : p1;
  const top0 = align === 0 ? leftTopLimit : align === 2 ? bottomLimit : p1;
  const top1 = align === 1 ? bottomLimit : align === 3 ? leftTopLimit : p1;
  const top2 = align === 0 ? bottomLimit : align === 2 ? leftTopLimit : 100 - p1;
  const top3 = align === 1 ? leftTopLimit : align === 3 ? bottomLimit : 100 - p1;

  return {
    left0: `${left0}%`,
    left2: `${left2}%`,
    left1: `${left1}%`,
    left3: `${left3}%`,
    top0: `${top0}%`,
    top1: `${top1}%`,
    top2: `${top2}%`,
    top3: `${top3}%`,
  };
}
