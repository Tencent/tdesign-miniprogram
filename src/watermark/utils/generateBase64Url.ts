import { systemInfo } from '../../common/utils';

export default function generateBase64Url(
  canvas,
  { width, height, gapX, gapY, offsetLeft, offsetTop, rotate, alpha, watermarkContent, lineSpace },
  onFinish,
) {
  const ctx = canvas.getContext('2d');
  if (!ctx) {
    console.warn('当前环境不支持Canvas, 无法绘制水印');
    onFinish('');
    return;
  }
  const ratio = systemInfo.pixelRatio || 1;
  const canvasWidth = (gapX + width) * ratio;
  const canvasHeight = (gapY + height) * ratio;

  canvas.width = canvasWidth;
  canvas.height = canvasHeight;

  ctx.translate(offsetLeft * ratio, offsetTop * ratio);
  ctx.rotate((Math.PI / 180) * Number(rotate));
  ctx.globalAlpha = alpha;

  const markWidth = width * ratio;
  const markHeight = height * ratio;

  ctx.fillStyle = 'transparent';
  ctx.fillRect(0, 0, markWidth, markHeight);

  const contents = Array.isArray(watermarkContent) ? watermarkContent : [{ ...watermarkContent }];

  let top = 0;
  contents.forEach((item) => {
    if (item.url) {
      const { url, isGrayscale = false } = item;
      item.top = top;
      top += height;
      const image = canvas.createImage();
      image.src = url;
      image.onload = () => {
        ctx.drawImage(image, 0, item.top * ratio, width * ratio, height * ratio);
        if (isGrayscale) {
          const imgData = ctx.getImageData(0, 0, ctx.canvas.width, ctx.canvas.height);
          const pixels = imgData.data;
          for (let i = 0; i < pixels.length; i += 4) {
            const lightness = (pixels[i] + pixels[i + 1] + pixels[i + 2]) / 3;
            pixels[i] = lightness;
            pixels[i + 1] = lightness;
            pixels[i + 2] = lightness;
          }
          ctx.putImageData(imgData, 0, 0);
        }
        onFinish(canvas.toDataURL());
      };
    } else if (item.text) {
      const {
        text,
        fontColor = 'rgba(0, 0, 0, 0.1)',
        fontSize = 16,
        fontFamily = undefined,
        fontWeight = 'normal',
      } = item;
      item.top = top;
      top += lineSpace;
      const markSize = Number(fontSize) * ratio;
      ctx.font = `normal normal ${fontWeight} ${markSize}px/${markHeight}px ${fontFamily}`;
      ctx.textAlign = 'start';
      ctx.textBaseline = 'top';
      ctx.fillStyle = fontColor;
      ctx.fillText(text, 0, item.top * ratio);
    }
  });
  onFinish(canvas.toDataURL());
}
