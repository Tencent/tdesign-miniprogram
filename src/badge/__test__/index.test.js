import simulate from 'miniprogram-simulate';
import path from 'path';
import { hex2Rgb } from '../../../test/utils/colors';

const shapes = ['circle', 'square', 'round', 'ribbon'];

const sizes = ['small', 'medium'];

describe('badge', () => {
  const badge = load(path.resolve(__dirname, `../badge`), 't-badge');

  it(`:base`, () => {
    const id = simulate.load({
      template: `<t-badge class="badge" dot content="测试"></t-badge>`,
      usingComponents: {
        't-badge': badge,
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    expect(comp.toJSON()).toMatchSnapshot();
  });

  it(':dot', () => {
    const id = simulate.load({
      template: `<t-badge class="badge" dot content="测试"></t-badge>`,
      usingComponents: {
        't-badge': badge,
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $dot = comp.querySelector('.badge >>> .t-badge--basic');
    expect($dot.dom.getAttribute('class').includes('t-badge--dot')).toBeTruthy();
  });

  it(':color', () => {
    const id = simulate.load({
      template: `<t-badge class="badge" color="{{color}}" dot content="测试"></t-badge>`,
      usingComponents: {
        't-badge': badge,
      },
      data: {
        color: '#332244',
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $dot = comp.querySelector('.badge >>> .t-badge--basic');
    expect($dot.dom.style.backgroundColor).toBe(hex2Rgb('#332244'));
  });

  it(':content', () => {
    const id = simulate.load({
      template: `<t-badge class="badge" content="{{content}}" dot></t-badge>`,
      usingComponents: {
        't-badge': badge,
      },
      data: {
        content: '消息提醒',
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $content = comp.querySelector('.badge >>> .t-badge__content-text');
    expect($content.dom.textContent).toBe('消息提醒');
  });

  it(':count', () => {
    const id = simulate.load({
      template: `<t-badge class="badge" count="{{count}}" content="测试"></t-badge>`,
      usingComponents: {
        't-badge': badge,
      },
      data: {
        count: 'new',
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $count = comp.querySelector('.badge >>> .t-badge--basic');
    expect($count.dom.textContent.trim()).toBe('new');
  });

  it(':max-count', () => {
    const id = simulate.load({
      template: `<t-badge class="badge" count="{{count}}" maxCount="{{maxCount}}" content="测试"></t-badge>`,
      usingComponents: {
        't-badge': badge,
      },
      data: {
        maxCount: 80,
        count: 90,
      },
    });
    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $count = comp.querySelector('.badge >>> .t-badge--basic');
    expect($count.dom.textContent.trim()).toBe('80+');
  });

  shapes.forEach((shape) => {
    it(`:shape ${shape}`, () => {
      const id = simulate.load({
        template: `<t-badge class="badge" count="{{count}}" shape="{{shape}}" content="测试"></t-badge>`,
        usingComponents: {
          't-badge': badge,
        },
        data: {
          shape,
          count: 90,
        },
      });
      const comp = simulate.render(id);
      comp.attach(document.createElement('parent-wrapper'));

      const $count = comp.querySelector('.badge >>> .t-badge--basic');
      expect($count.dom.getAttribute('class').includes(`t-badge--${shape}`)).toBeTruthy();
    });
  });

  sizes.forEach((size) => {
    it(`:size ${size}`, () => {
      const id = simulate.load({
        template: `<t-badge class="badge" dot size="{{size}}" content="测试"></t-badge>`,
        usingComponents: {
          't-badge': badge,
        },
        data: {
          size,
        },
      });
      const comp = simulate.render(id);
      comp.attach(document.createElement('parent-wrapper'));

      const $count = comp.querySelector('.badge >>> .t-badge--basic');
      if (size === 'small') {
        expect($count.dom.getAttribute('class').includes(`t-badge--${size}`)).toBeTruthy();
      } else {
        expect($count.dom.getAttribute('class').includes(`t-badge--${size}`)).not.toBeTruthy();
      }
    });
  });

  it(':show-zero', async () => {
    const id = simulate.load({
      template: `<t-badge class="badge" count="{{count}}" showZero="{{showZero}}" content="测试"></t-badge>`,
      usingComponents: {
        't-badge': badge,
      },
      data: {
        showZero: false,
        count: 0,
      },
    });

    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $count = comp.querySelector('.badge >>> .t-badge--basic');
    expect($count).toBeUndefined();

    comp.setData({ showZero: true });

    const $count2 = comp.querySelector('.badge >>> .t-badge--basic');
    expect($count2.dom.textContent.trim()).toBe('0');
  });

  it(':offset string with unit', async () => {
    const id = simulate.load({
      template: `<t-badge class="badge" count="{{count}}" offset="{{offset}}" content="测试"></t-badge>`,
      usingComponents: {
        't-badge': badge,
      },
      data: {
        count: 9,
        offset: ['15px', '20em'],
      },
    });

    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $count = comp.querySelector('.badge >>> .t-badge--basic');
    expect($count.dom.style.top).toBe('15px');
    expect($count.dom.style.right).toBe('20em');
  });

  it(':offset number without unit', async () => {
    const id = simulate.load({
      template: `<t-badge class="badge" count="{{count}}" offset="{{offset}}" content="测试"></t-badge>`,
      usingComponents: {
        't-badge': badge,
      },
      data: {
        count: 9,
        offset: [16, 29],
      },
    });

    const comp = simulate.render(id);
    comp.attach(document.createElement('parent-wrapper'));

    const $count = comp.querySelector('.badge >>> .t-badge--basic');
    expect($count.dom.style.top).toBe('16px');
    expect($count.dom.style.right).toBe('29px');
  });
});
