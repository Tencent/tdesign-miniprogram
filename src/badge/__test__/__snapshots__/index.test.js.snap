// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`badge :base 1`] = `
<main>
  <t-badge
    class="badge"
  >
    <wx-view
      class="t-badge  t-class"
      style=""
    >
      <wx-view
        class="t-badge__content t-class-content"
      >
        <wx-text
          class="t-badge__content-text"
        >
          测试
        </wx-text>
      </wx-view>
      <wx-view
        class="t-badge--basic t-badge--dot  t-badge--circle  t-has-count t-class-count"
        style=""
      >
        
  
      </wx-view>
    </wx-view>
  </t-badge>
</main>
`;
